### PR TITLE
Port the ``EvolvedOperatorAnsatz`` from Nature

### PR DESCRIPTION
**Pull request by Cryoris**
### Summary

Port the ``EvolvedOperatorAnsatz`` from Nature to core.

Related to Qiskit/qiskit-nature#117.

### Details and comments

It was initially located in Nature as we needed it for the UCCSD circuits and couldn't wait for a Terra release. However, it should live in Terra as it can be the base for other library circuits, such as QAOA (which is included in this PR).

Two lines in the tests have been modified since they relied on a particular random array but with this PR and the changed code the random array changed. I changed the test criterion from a comparison to the final parameters to the final loss.

### CODE REVIEW

File: qiskit/circuit/library/evolved_operator_ansatz.py [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+
+    def __init__(
+        self,
+        operators: Optional[
+            Union["OperatorBase", QuantumCircuit, List[Union["OperatorBase", QuantumCircuit]]]
+        ] = None,
+        reps: int = 1,
+        evolution: Optional["EvolutionBase"] = None,
+        insert_barriers: bool = False,
+        name: str = "EvolvedOps",
```
Comment (mrossinek) [inline qiskit/circuit/library/evolved_operator_ansatz.py:36]: Just a quick question: why not just `"EvolvedOp"`?

Comment (Cryoris) [reply qiskit/circuit/library/evolved_operator_ansatz.py:36]: because it can be a list of operators 😄

**Review by woodsp-ibm — APPROVED** [review]

**Review by woodsp-ibm — APPROVED** [review]
I guess the mix of typehints in the signature and docstring was unavoidable in someway in order to get the documentation built. It would be nicer of course not to have such a mix - maybe things can be improved later.

**Comment by Cryoris** [comment]
Yeah that's a bit annoying (as you can see from the last 4 Sphinx commits 😉). I'm not sure how we can fix this in the future since these are cyclic imports. Maybe Python 3.7s typing system can fix it?

File: qiskit/circuit/library/n_local/qaoa_ansatz.py [hunk 2]
```diff
@@ -38,7 +34,7 @@ def __init__(
         reps: int = 1,
         initial_state: Optional[QuantumCircuit] = None,
         mixer_operator=None,
-        name: str = "qaoa",
+        name: str = "QAOA",
```
Comment (mtreinish) [inline qiskit/circuit/library/n_local/qaoa_ansatz.py:37]: Is this a potentially breaking change for people if they're matching on the default name?

**Review by mtreinish — APPROVED** [review]
LGTM, inline comment. But it's not a blocker if it is an issue we can fix in a follow up before the 0.18.0 release.

**Comment by t-imamichi** [comment]
@Cryoris Why do you need to change the parameter order? It causes a unit test failure at qiskit-optimization. See #6424.

**Comment by ikkoham** [comment]
In addition, this PR may have forgotten to set the bounds.
In the past, we set bounds [here](https://github.com/Qiskit/qiskit-terra/pull/6396/files#diff-2dcd521524e5e468f6d60138c49f247a683e4d5cc275c94ba88c3aa70113a758L165-L175).
This changes the [test result](https://github.com/Qiskit/qiskit-terra/pull/6396/files#diff-97866335b75116c0c55610099daaa8deb8014daf8a615b45b991daefa79bcb66L300).
```
Before:  x: array([-0.56 ,  0.315])
After: y: array([-0.564, -5.96 ])
```

**Pull request merged**